src/hashset.c

```diff
@@ -0,0 +1,1700 @@
+    int emit_ref = (flags & HASHSET_SCAN_EMIT_REF);
+    int single_step = (flags & HASHSET_SCAN_SINGLE_STEP);
+
+    /* If any element that hashes to the current bucket may have been inserted
+     * in another bucket due to probing, we need to continue to cover the whole
+     * probe sequence in the same scan cycle. Otherwise we may miss those
+     * elements if they are rehashed before the next scan call. */
+    int in_probe_sequence = 0;
+
+    /* When the cursor reaches zero, may need to continue scanning and advancing
+     * the cursor until the probing chain ends, but when we stop, we return 0 to
+     * indicate that the full scan is completed. */
+    int cursor_passed_zero = 0;
+
+    /* Mask the start cursor to the bigger of the tables, so we can detect if we
+     * come back to the start cursor and break the loop. It can happen if enough
+     * tombstones (in both tables while rehashing) make us continue scanning. */
+    cursor = cursor & (expToMask(s->bucket_exp[0]) | expToMask(s->bucket_exp[1]));
+    size_t start_cursor = cursor;
+    do {
+        in_probe_sequence = 0; /* Set to 1 if an ever-full bucket is scanned. */
+        if (!hashsetIsRehashing(s)) {
+            /* Emit elements at the cursor index. */
+            size_t mask = expToMask(s->bucket_exp[0]);
+            bucket *b = &s->tables[0][cursor & mask];
+            int pos;
+            for (pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+                if (isPositionFilled(b, pos)) {
+                    void *emit = emit_ref ? &b->elements[pos] : b->elements[pos];
+                    fn(privdata, emit);
+                }
+            }
+
+            /* Do we need to continue scanning? */
+            in_probe_sequence |= b->everfull;
+
+            /* Advance cursor. */
+            cursor = nextCursor(cursor, mask);
+        } else {
+            int table_small, table_large;
+            if (s->bucket_exp[0] <= s->bucket_exp[1]) {
+                table_small = 0;
+                table_large = 1;
+            } else {
+                table_small = 1;
+                table_large = 0;
+            }
+
+            size_t mask_small = expToMask(s->bucket_exp[table_small]);
+            size_t mask_large = expToMask(s->bucket_exp[table_large]);
+
+            /* Emit elements in the smaller table, if this bucket hasn't already
+             * been rehashed. */
+            if (table_small == 0 && !cursorIsLessThan(cursor, s->rehash_idx)) {
+                bucket *b = &s->tables[table_small][cursor & mask_small];
+                if (b->presence) {
+                    for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+                        if (isPositionFilled(b, pos)) {
+                            void *emit = emit_ref ? &b->elements[pos] : b->elements[pos];
+                            fn(privdata, emit);
+                        }
+                    }
+                }
+                in_probe_sequence |= b->everfull;
+            }
```

Comment:
I just realized this condition I've been staring at is wrong. If table_small == 1 (the new table) it won't look at the bucket ever.